### PR TITLE
Alerting: validate alert condition on saving rule

### LINKED ISSUE #61832: Block saving alert rules with invalid alert conditions
When an alert rule is created (or edited) users need to set an alert condition for it to be evaluated. These alert conditions must result in a number (1 or 0) or several for multidimensional alerts, but they can't consist of a timeseries result.

The UI warns you in these cases

image

image

But even if there's a warning message, it still allows you to save the form resulting in an alert that will never fire:

### PR DESCRIPTION
**Pull request by VikaCep**
**What is this feature?**

This PR prevents alert rules to be created with invalid conditions, such as when the result of the query/expression results in timeseries data.

**Why do we need this feature?**

We need to improve the alert rule creation process, making errors more obvious to the user.

**Who is this feature for?**

All users.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/61832

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx [hunk 1]
```diff
@@ -104,6 +105,30 @@ export const QueryAndExpressionsStep: FC<Props> = ({ editingExistingRule }) => {
...
+  useEffect(() => {
+    const currentCondition = getValues('condition');
+
+    if (!currentCondition) {
+      return;
+    }
+
+    const error = errorFromSeries(panelData[currentCondition]?.series || []);
+    onDataChange(error?.message || '');
+  }, [panelData, getValues, onDataChange]);
```
Comment (VikaCep) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx:117]: The reason I'm using this `useEffect` with `panelData` is to be able to check if the condition is valid **once** the data has been updated as a result of `runQueries()`. This happens after either setting a new condition or making changes to the query of the current one.

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM! nice one!  🎉
Adding a non blocking question 👀

**Comment by soniaAguilarPeiron** [comment]
I have a question regarding no data : should we consider that the condition is valid in this case?
Screenshot 2023-01-24 at 10 02 03

**Comment by VikaCep** [comment]
> I have a question regarding no data : should we consider that the condition is valid in this case?

@soniaAguilarPeiron I assumed it's a valid scenario since that's something that may depend on the interval. The condition could be valid even if at the given interval there's no data. I think it's kind of hard to know whether the query/expression will return data at a different point in time so I didn't wanna mark it as something wrong. But I can change it if you disagree.

**Comment by VikaCep** [comment]
After digging a little bit deeper I noticed when the result of the query is empty (for example when we filter by a metric that doesn't return any data for the given interval), we get an empty `series` array. The same thing happens when there is no metrics or label filters assigned to the query, as per your screenshot, so it's tricky to tell the real reason why the query returns no data. Also, I think we should be careful to tell whether a query is invalid when it lacks a metric/filter because its format depends on the datasource type; not all queries expect an input like a PromQL query. I think we could go with this solution as is (since previous to this PR empty queries were considered valid conditions) and if there's need to further restrict the validation we can do that in a separate PR.

**Pull request merged**